File: 0105-construct-binary-tree-from-preorder-and-inorder-traversal/0105-construct-binary-tree-from-preorder-and-inorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        Map<Integer, Integer> map = new HashMap<>();
        for (int i = 0; i < inorder.length; i++) {
            map.put(inorder[i], i);
        }
        
        return helper(preorder, 0, preorder.length - 1, map, 0, inorder.length - 1);
    }

    private TreeNode helper(int[] preorder, int preLeft, int preRight, Map<Integer, Integer> inorder, int inLeft, int inRight) {
        if (preLeft > preRight || inLeft > inRight) {
            return null;
        }
        
        TreeNode cur = new TreeNode(preorder[preLeft]);

        int newInLeft = inorder.get(cur.val);
        int leftSubtreeSize = newInLeft - inLeft;

        cur.left = helper(preorder, preLeft + 1, preLeft + leftSubtreeSize, inorder, inLeft, newInLeft - 1);
        cur.right = helper(preorder, preLeft + leftSubtreeSize + 1, preRight, inorder, newInLeft + 1, inRight);

        return cur;
    }
}

File: 0105-construct-binary-tree-from-preorder-and-inorder-traversal/0105-construct-binary-tree-from-preorder-and-inorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        unordered_map<int, int> inorderIndexMap;
        for (int i = 0; i < inorder.size(); i++) {
            inorderIndexMap[inorder[i]] = i;
        }

        int preorderIndex = 0;
        return helper(preorder, preorderIndex, inorderIndexMap, 0, inorder.size() - 1);
    }

private:
    TreeNode* helper(vector<int>& preorder, int& preorderIndex, 
                     unordered_map<int, int>& inorderIndexMap, int inLeft, int inRight) {
        if (inRight < inLeft) {
            return nullptr;
        }

        TreeNode* cur = new TreeNode(preorder[preorderIndex++]);
        cur->left = helper(preorder, preorderIndex, inorderIndexMap, inLeft, inorderIndexMap[cur->val] - 1);
        cur->right = helper(preorder, preorderIndex, inorderIndexMap, inorderIndexMap[cur->val] + 1, inRight);

        return cur;     
    }
};